File: web/src/krate/kfile/dto.ts
import type { TID } from "@/lib/id_util";
import type { Varchar } from "@/lib/types";
import type { InlineKFile, KFileMeta } from "./po";

export type KFileUploadRsp = {
  kfile?: KFileMeta;
  finished: boolean;
};

export type InlineKFileUploadReq = {
  meta_id: Varchar<100>;
  otid: TID;
  res: InlineKFile;
  archor_intervals: number;
  filename?: Varchar<1024>;
  content_type: Varchar<200>;
  binaryp: boolean;
};

export type InlineKFileUploadRsp = {
  true_sid: Varchar<100>;
};

export type KfileMetaFetchReqId = { Otid: TID } | { Id: string };

export type InlineKFileDownloadReq = {
  req_id: KfileMetaFetchReqId;
  with_omit?: boolean;
};

export type InlineKFileDownloadRsp = {
  meta?: KFileMeta;
  file?: InlineKFile;
};

export type KfileMetaFetchReq = {
  req_id: KfileMetaFetchReqId;
  history_and_archor?: boolean;
};

export type KfileMetaFetchRsp = {
  meta?: KFileMeta;
};

export type KfileHistoryListReq = {
  otid: TID;
};

export type KfileHistoryVersion = {
  tid: TID;
};

export type KfileHistoryListRsp = {
  versions: KfileHistoryVersion[];
};

export type KfileHistoryFetchReq = {
  otid: TID;
  tid: TID;
};

export type KfileHistoryFetchRsp = {
  meta?: KFileMeta;
  file?: InlineKFile;
};

export type KfileHistoryApplyReq = {
  otid: TID;
  tid: TID;
};

export type KfileHistoryApplyRsp = {
  meta?: KFileMeta;
  file?: InlineKFile;
};

export type InlineKFileUploadDirectlyReq = {
  file: InlineKFile;
};
export type InlineKFileUploadDirectlyRsp = object;
export type InlineKFileDownloadBySidReq = {
  sid: Varchar<100>;
};
export type InlineKFileDownloadBySidRsp = {
  file?: InlineKFile;
};

export type KfileAssetChunkUploadReq = {
  upload_id: string;
  meta_id: string;
  otid: number;
  filename: string;
  chunk_no: number;
  total_chunks: number;
  chunk: Blob;
  last_modified: number;
  filesize: number;
  content_type: string;
  db_store?: boolean;
  binaryp: boolean;
};

File: lib/backend/src/krate/kfile/dto.rs
use super::*;
use axum::body::Bytes;
use axum_typed_multipart::{FieldData, TryFromMultipart};
use chin_sql::{str_type::Varchar, time_type::TID};
use serde::{Deserialize, Serialize};

#[derive(TryFromMultipart)]
pub struct KfileAssetChunkUploadReq {
    pub upload_id: String,
    pub meta_id: String,
    pub otid: i64,
    pub filename: String,
    pub chunk_no: usize,
    pub total_chunks: usize,
    pub chunk: FieldData<Bytes>,
    pub last_modified: i64,
    pub filesize: i64,
    pub content_type: String,
    pub db_store: Option<bool>,
    pub binaryp: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KFileUploadRsp {
    pub kfile: Option<KFileMeta>,
    pub finished: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InlineKFileUploadReq {
    pub meta_id: Varchar<100>,
    pub otid: TID,
    pub res: InlineKFile,
    /// archor interval second.
    pub archor_intervals: i64,
    pub filename: Option<Varchar<1024>>,
    pub content_type: Varchar<200>,
    pub binaryp: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InlineKFileUploadRsp {
    pub true_sid: Varchar<100>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InlineKFileDownloadReq {
    pub req_id: KfileMetaFetchReqId,
    pub with_omit: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InlineKFileDownloadRsp {
    pub meta: Option<KFileMeta>,
    pub file: Option<InlineKFile>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum KfileMetaFetchReqId {
    Otid(TID),
    Id(Varchar<100>),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KfileMetaFetchReq {
    pub req_id: KfileMetaFetchReqId,
    pub history_and_archor: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KfileMetaFetchRsp {
    pub meta: Option<KFileMeta>,
}

pub const KFILE_ASSET_UPLOAD_BY_SID: &str = "/api/v1/kfile-asset-upload-by-sid";

#[test]
fn tst() {
    let c = KfileMetaFetchReqId::Otid(100.try_into().unwrap());
    println!("{}", serde_json::to_string_pretty(&c).unwrap());
}
